kong/plugins/bot-detection/handler.lua

```diff
@@ -19,10 +44,79 @@ local function get_user_agent()
   return user_agent
 end
 
+local function examine_agent(user_agent, conf)
+  user_agent = strip(user_agent)
+
+  if conf.whitelist then
+    for _, rule in ipairs(conf.whitelist) do
+      if re_match(user_agent, rule) then
+        return MATCH_WHITELIST
+      end
+    end
+  end
+
+  if conf.blacklist then
+    for _, rule in ipairs(conf.blacklist) do
+      if re_match(user_agent, rule) then
+        return MATCH_BLACKLIST
+      end
+    end
+  end
+
+  for _, rule in ipairs(rules.bots) do
+    if re_match(user_agent, rule) then
+      return MATCH_BOT
+    end
+  end
+
+  return MATCH_EMPTY
+end
+
 function BotDetectionHandler:new()
   BotDetectionHandler.super.new(self, "bot-detection")
 end
 
+function BotDetectionHandler:init_worker()
+  local singletons    = require "kong.singletons"
+  local worker_events = singletons.worker_events
+  local dao_factory   = singletons.dao
+
+  -- load our existing plugins to create our cache spaces
+  local plugins, err = dao_factory.plugins:find_all({
+    name = "bot-detection",
+  })
+  if err then
+    ngx.log(ngx.ERR, "err in fetching plugins: ", err)
+  end
+
+  for i = 1, #plugins do
+    ua_cache[plugins[i].api_id] = {}
+  end
+
+  -- catch updates and tell each worker to adjust their cache
+  -- accordingly. for create/updates we create/purge the cache
+  -- by simply creating a new lua table; for deletions, as we no
+  -- longer need this cache data, we just remove it
+  worker_events.register(function(data)
+    if data.entity.name == "bot-detection" then
+      worker_events.post("bot-detection-invalidate", data.operation,
```

Comment:
Can you help me understand this further? I think I see the behavior you are describing. Consider the following debug log patch applied:

```
$ git diff kong/plugins/bot-detection/handler.lua 
diff --git a/kong/plugins/bot-detection/handler.lua b/kong/plugins/bot-detection/handler.lua
index f6a08fd..52f9d71 100644
--- a/kong/plugins/bot-detection/handler.lua
+++ b/kong/plugins/bot-detection/handler.lua
@@ -105,14 +105,17 @@ function BotDetectionHandler:init_worker()
   end, "crud", "plugins")
 
   worker_events.register(function(entity)
+    ngx.log(ngx.NOTICE, "create received on pid ", ngx.worker.pid())
     ua_cache[entity.api_id] = {}
   end, "bot-detection-invalidate", "create")
 
   worker_events.register(function(entity)
+    ngx.log(ngx.NOTICE, "update received on pid ", ngx.worker.pid())
     clear_tab(ua_cache[entity.api_id])
   end, "bot-detection-invalidate", "update")
 
   worker_events.register(function(entity)
+    ngx.log(ngx.NOTICE, "delete received on pid ", ngx.worker.pid())
     ua_cache[entity.api_id] = nil
   end, "bot-detection-invalidate", "delete")
 end
```

i see the following in the error logs on create:

```
2017/07/08 04:48:23 [notice] 4305#0: *55 [lua] handler.lua:108: create received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "POST /apis/test/plugins HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:23 [info] 4305#0: *55 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:23 [notice] 4303#0: *61 [lua] handler.lua:108: create received on pid 4303, context: ngx.timer
2017/07/08 04:48:23 [notice] 4302#0: *62 [lua] handler.lua:108: create received on pid 4302, context: ngx.timer
2017/07/08 04:48:23 [notice] 4304#0: *63 [lua] handler.lua:108: create received on pid 4304, context: ngx.timer
```

on update:

```
2017/07/08 04:48:38 [notice] 4305#0: *139 [lua] handler.lua:113: update received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "PATCH /apis/test/plugins/42b1b1b6-ecd0-44e1-b02b-e4e89dc2e3bf HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:38 [info] 4305#0: *139 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:39 [notice] 4304#0: *141 [lua] handler.lua:113: update received on pid 4304, context: ngx.timer
2017/07/08 04:48:39 [notice] 4303#0: *142 [lua] handler.lua:113: update received on pid 4303, context: ngx.timer
2017/07/08 04:48:39 [notice] 4302#0: *143 [lua] handler.lua:113: update received on pid 4302, context: ngx.timer
```

and delete:

```
2017/07/08 04:48:45 [notice] 4305#0: *173 [lua] handler.lua:118: delete received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "DELETE /apis/test/plugins/42b1b1b6-ecd0-44e1-b02b-e4e89dc2e3bf HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:45 [info] 4305#0: *173 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:46 [notice] 4304#0: *174 [lua] handler.lua:118: delete received on pid 4304, context: ngx.timer
2017/07/08 04:48:46 [notice] 4303#0: *175 [lua] handler.lua:118: delete received on pid 4303, context: ngx.timer
2017/07/08 04:48:46 [notice] 4302#0: *176 [lua] handler.lua:118: delete received on pid 4302, context: ngx.timer
```

for reference, the full debug log following restart:

```
2017/07/08 04:48:23 [debug] 4305#0: *55 [lua] init.lua:139: Loading API endpoints for plugin: key-auth
2017/07/08 04:48:23 [debug] 4305#0: *55 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:48:23 [debug] 4305#0: *55 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:48:23 [notice] 4305#0: *55 [lua] handler.lua:108: create received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "POST /apis/test/plugins HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:23 [info] 4305#0: *55 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:23 [notice] 4303#0: *61 [lua] handler.lua:108: create received on pid 4303, context: ngx.timer
2017/07/08 04:48:23 [notice] 4302#0: *62 [lua] handler.lua:108: create received on pid 4302, context: ngx.timer
2017/07/08 04:48:23 [notice] 4304#0: *63 [lua] handler.lua:108: create received on pid 4304, context: ngx.timer
2017/07/08 04:48:26 [info] 4305#0: *77 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:27 [debug] 4303#0: *78 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489292.382 to: 1499489307.388
2017/07/08 04:48:32 [debug] 4302#0: *103 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489307.387 to: 1499489312.389
2017/07/08 04:48:37 [debug] 4304#0: *128 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489307.387 to: 1499489317.397
2017/07/08 04:48:38 [debug] 4305#0: *139 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:48:38 [debug] 4305#0: *139 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:48:38 [debug] 4305#0: *139 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:48:38 [debug] 4305#0: *139 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:48:38 [notice] 4305#0: *139 [lua] handler.lua:113: update received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "PATCH /apis/test/plugins/42b1b1b6-ecd0-44e1-b02b-e4e89dc2e3bf HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:38 [info] 4305#0: *139 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:39 [notice] 4304#0: *141 [lua] handler.lua:113: update received on pid 4304, context: ngx.timer
2017/07/08 04:48:39 [notice] 4303#0: *142 [lua] handler.lua:113: update received on pid 4303, context: ngx.timer
2017/07/08 04:48:39 [notice] 4302#0: *143 [lua] handler.lua:113: update received on pid 4302, context: ngx.timer
2017/07/08 04:48:42 [debug] 4302#0: *153 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489307.387 to: 1499489322.399
2017/07/08 04:48:45 [debug] 4305#0: *173 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:48:45 [debug] 4305#0: *173 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:48:45 [notice] 4305#0: *173 [lua] handler.lua:118: delete received on pid 4305, client: 127.0.0.1, server: kong_admin, request: "DELETE /apis/test/plugins/42b1b1b6-ecd0-44e1-b02b-e4e89dc2e3bf HTTP/1.1", host: "localhost:8001"
2017/07/08 04:48:45 [info] 4305#0: *173 client 127.0.0.1 closed keepalive connection
2017/07/08 04:48:46 [notice] 4304#0: *174 [lua] handler.lua:118: delete received on pid 4304, context: ngx.timer
2017/07/08 04:48:46 [notice] 4303#0: *175 [lua] handler.lua:118: delete received on pid 4303, context: ngx.timer
2017/07/08 04:48:46 [notice] 4302#0: *176 [lua] handler.lua:118: delete received on pid 4302, context: ngx.timer
2017/07/08 04:48:47 [debug] 4303#0: *178 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489322.398 to: 1499489327.401
2017/07/08 04:48:52 [debug] 4303#0: *202 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489327.4 to: 1499489332.401
2017/07/08 04:48:57 [debug] 4304#0: *226 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489327.4 to: 1499489337.403
```

if we change the `worker_event.post` to `worker_events.post_local` , not all workers catch this update:

```
2017/07/08 04:54:50 [debug] 4618#0: *145 [lua] init.lua:139: Loading API endpoints for plugin: key-auth
2017/07/08 04:54:50 [debug] 4618#0: *145 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:54:50 [debug] 4618#0: *145 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:54:50 [notice] 4618#0: *145 [lua] handler.lua:109: create received on pid 4618, client: 127.0.0.1, server: kong_admin, request: "POST /apis/test/plugins HTTP/1.1", host: "localhost:8001"
2017/07/08 04:54:50 [info] 4618#0: *145 client 127.0.0.1 closed keepalive connection
2017/07/08 04:54:51 [debug] 4615#0: *151 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489661.125 to: 1499489691.141
2017/07/08 04:54:56 [debug] 4617#0: *175 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489691.14 to: 1499489696.142
2017/07/08 04:55:01 [debug] 4616#0: *199 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489691.14 to: 1499489701.144
2017/07/08 04:55:01 [debug] 4618#0: *207 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:55:01 [debug] 4618#0: *207 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:55:01 [debug] 4618#0: *207 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:55:01 [debug] 4618#0: *207 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:55:01 [notice] 4618#0: *207 [lua] handler.lua:114: update received on pid 4618, client: 127.0.0.1, server: kong_admin, request: "PATCH /apis/test/plugins/34ca6fdd-3078-4b17-8e04-96f758a8a2df HTTP/1.1", host: "localhost:8001"
2017/07/08 04:55:01 [info] 4618#0: *207 client 127.0.0.1 closed keepalive connection
2017/07/08 04:55:05 [debug] 4618#0: *223 [lua] cache.lua:128: invalidate_local(): [DB cache] invalidating (local): 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:'
2017/07/08 04:55:05 [debug] 4618#0: *223 [lua] cache.lua:149: invalidate(): [DB cache] broadcasting (cluster) invalidation for key: 'plugins:bot-detection:4006a139-8f1f-49a3-a7d7-03b1654075e6:' with nbf: 'none'
2017/07/08 04:55:05 [notice] 4618#0: *223 [lua] handler.lua:119: delete received on pid 4618, client: 127.0.0.1, server: kong_admin, request: "DELETE /apis/test/plugins/34ca6fdd-3078-4b17-8e04-96f758a8a2df HTTP/1.1", host: "localhost:8001"
2017/07/08 04:55:05 [info] 4618#0: *223 client 127.0.0.1 closed keepalive connection
2017/07/08 04:55:06 [debug] 4617#0: *225 [lua] cluster_events.lua:231: [cluster_events] polling events from: 1499489691.14 to: 1499489706.144
```

so with this implementation we don't have the desired effect (updating all workers on the node when a crud event is handled). let me know your thoughts here.